Add option to display OSD menus over a solid gray background

etracer65 opened the pull request:
Improves menu readability by changing the background from a transparent display of the camera image to a static opaque gray background.

The behavior is controlled with the `osd_menu_background` parameter which defaults to `TRANSPARENT` to preserve the previous behavior. Other opaque options are available:
```
osd_menu_background = TRANSPARENT
Allowed values: TRANSPARENT, BLACK, GRAY, LIGHT_GRAY
```

The background setting is available in the CMS OSD menu and the user can cycle through the various options with the display updating in real-time.

Currently only the onboard MAX7456-based OSD is supported, but the implementation adds `displayPort` support so it can easily be extended to other OSD devices if those manufacturers want to add support. Also can be extended to other background types (like colors, varying transparency, etc.) for future device support.

Makes use of the built-in MAX7456 feature to display all transparent pixels as "gray". The MAX7456 display area seems to be a few scan lines smaller than the actual camera video image so it's normal for some of the camera image to "leak" at the top/bottom of the display. The OSD display area can be adjusted up/down using the `vcd_v_offset` setting if desired.

[betaflight_4.3.0_STM32F411_89c9609.zip](https://github.com/betaflight/betaflight/files/5761926/betaflight_4.3.0_STM32F411_89c9609.zip)
[betaflight_4.3.0_STM32F405_89c9609.zip](https://github.com/betaflight/betaflight/files/5761927/betaflight_4.3.0_STM32F405_89c9609.zip)
[betaflight_4.3.0_STM32F7X2_89c9609.zip](https://github.com/betaflight/betaflight/files/5761928/betaflight_4.3.0_STM32F7X2_89c9609.zip)
[betaflight_4.3.0_STM32F745_89c9609.zip](https://github.com/betaflight/betaflight/files/5761929/betaflight_4.3.0_STM32F745_89c9609.zip)

etracer65 commented:
I was only able to test with a NTSC camera so it would be good to have additional testing with PAL.

Igorshp commented:
Awesome, love it. This is way cleaner way to do this.

There is a bug though.  If you enter the menu stick command again while the menu is open, it closes the menu but does *not* clear the gray background.

Asizon commented:
This feature looks so usefull, thanks for implement it @etracer65 , and thanks @Igorshp for the idea, I see some days ago. I wonder if this affects the cam control osd when adjusting parameters?

Igorshp commented:
> This feature looks so usefull, thanks for implement it @etracer65 , and thanks @Igorshp for the idea, I see some days ago and I wonder if this affects the cam control osd when adjusting parameters?

Yeah, it's definitely a lot more comfortable to read the menu like this, though personal preference would be black background. I like that @etracer65 left an option for user choice.

Cam control is not affected as it's rendered on the camera and not on the flight controller osd chip. Would've been great though.

etracer65 commented:
> Awesome, love it. This is much better and way cleaner.
> 
> There is a bug though. If you enter the menu stick command while the menu is open, it closes the menu but does **not** clear the gray background.

I will fix the display resetting, but note that you actually aren't exiting the CMS. Doing the stick command to invoke the CMS again actually is used to rotate the CMS to the next available displayPort. So the menus are switching from the onboard MAX7456 to another displayPort device. You can see this if you do the sequence you described and then try to arm. You'll get a `CMS` arming disabled warning. I think there is a separate unrelated bug in that another displayPort device (probably MSP displayPort) is registering when it shouldn't. Or there's a problem with the "next" displayPort logic. But that's separate from this PR. But yes it's a bug in this PR that it should remove the dark background when switching to the next displayPort and I'll fix.

etracer65 commented:
> This feature looks so usefull, thanks for implement it @etracer65 , and thanks @Igorshp for the idea, I see some days ago and I wonder if this affects the cam control osd when adjusting parameters?

This PR doesn't affect camera control at all. As noted, that "OSD" is generated by the camera so we have no control over it. Also you probably really wouldn't want this as you're adjusting the camera because in many cases you want to actually see the camera image and the effect of what you're adjusting. Lastly we "blindly" invoke the cameras OSD menu. There is no feedback as to whether it's actually in the menu or if/when it exits so there's no way we would know when to turn on/off the dark background (if we could).

etracer65 commented:
Updated to fix the background not clearing when switching active displayPorts.

Added more "color options": `TRANSPARENT, BLACK, GRAY, LIGHT_GRAY`
So now people can choose one that best fits their needs. The `LIGHT_GRAY` option would be good for fonts that are mostly black.

Added to the CMS under the OSD menu to enable changing the background with real-time feedback.

Test builds in the PR description have been updated.

etracer65 commented:
> and thanks @Igorshp for the idea

I'm confused by this. I did this PR based on a request in the Facebook user group and some earlier work on MAX7456 emulation. Not sure how @Igorshp was involved. The FB post is here: https://www.facebook.com/groups/betaflight/permalink/1355754284774321/

Asizon commented:
Yes sorry @etracer65 it was my confusion, definitelly is a coincidence that i have seen this in their repo. In the other hand good to know that this feature doesnt afect to cam control. Thankyou @etracer65

etracer65 commented:
> What about adding to MSP??

We only add new features to MSP one release after the feature was implemented (so in 4.4). This gives time for the feature to mature and become stable. It's extremely difficult to change the MSP messages once they've been updated so we don't do this initially.

etracer65 commented:
Also important to point out that there are some cloned MAX7456 chip variants out there that have incomplete implementations. So it's possible this may not work on some flight controllers using bad clones. But nothing we can do about it. Should just fall back to normal transparent mode.

Also I know there are some MAX7456 "emulation" implementations that probably won't support this initially as we've never used this MAX7456 register before (so they probably didn't implement it).

Asizon commented:
@etracer65 tested with MatekF411RX(Betafpvf411 board). It seems that it cannot occupy the entire screen and there is a line underneath without covering:

![osdbackground](https://user-images.githubusercontent.com/43983086/103620094-30778e00-4f33-11eb-8ec4-6d9ba13f046e.jpg)

etracer65 commented:
> @etracer65 tested with MatekF411RX(Betafpvf411 board), camera i think is NTSC. It seems that it cannot occupy the entire screen and there is a line underneath without covering:

Yes, this is normal and noted in the PR description.

> The MAX7456 display area seems to be a few scan lines smaller than the actual camera video image so it's normal for some of the camera image to "leak" at the top/bottom of the display. The OSD display area can be adjusted up/down using the `vcd_v_offset` setting if desired.

mikeller commented:
Tested with PAL, works fine.
PAL has a bar of video image showing at the bottom as well.

etracer65 commented:
Rebased.

etracer65 commented:
@fiam You might want to consider adding support for this in FrskyOSD. It's implemented as a displayPort extension so you would just need to add the method and support in the FrskyOSD firmware. The list of options for the background could potentially be extended as well for things like various levels of transparency, etc. if the hardware is capable.
